### PR TITLE
zabbix_host.present: many fixes

### LINKED ISSUE #53838: zabbix_host module does not process additional host properties
### Description of Issue
I try to create a new Zabbix host with saltstack.
The description of Zabbix_host.present stated this:
`This function accepts all standard host properties: keyword argument names differ depending on your zabbix version, see:https://www.zabbix.com/documentation/2.4/manual/api/reference/host/object#host`
I have added for example  tls keywords to the  Zabbix_host.present  command in the sls file:

### Setup
```
zabbix_create_host:
  zabbix_host.present:
    - host: {{ salt['grains.get']('nodename') }}
    - groups:
      - 20
    - interfaces:
        - default:
            - ip: '0.0.0.0'
            - dns: ''
            - type: 'agent'
            - port: '10050'
    - inventory_mode: 1
    - tls_connect: 2
    - tls_accept: 2
    - tls_psk_identity: PSK_{{ salt['grains.get']('nodename') }}
    - tls_psk: {{ salt['pillar.get']('zabbix:psk') }}
```
### Steps to Reproduce Issue
minion log file:
```
data: {"jsonrpc": "2.0", "params": {"interfaces": [{"ip": "0.0.0.0", "useip": "1", "dns": "", "main": "1", "type": "1", "port": "10050"}], "proxy_hostid": "0", "host": "<hostname>", "groups": [{"groupid": 20}], "inventory": {}}, "id": 0, "auth": "<auth_token>", "method": "host.create"}
```

### Versions Report
All versions affected

### LINKED ISSUE #58602: [BUG] visible_name is ignored by zabbix_host.present
**Description**
The optional parameter visible_name is ignored by zabbix_host.present. This happens because it's not passed to host_create in zabbix.py execution module:
```
369         host_create = __salt__["zabbix.host_create"](
370             host,
371             groups,
372             interfaces_formated,
373             proxy_hostid=proxy_hostid,
374             inventory=new_inventory,
375             **connection_args
376         )
```

**Setup**
Slackware machines running Salt 3001 managing hosts in a Zabbix 5.0 server, also running on Slackware. The rendered state being applied is:
```
create-zabbix-host-bat.mylab:
  zabbix_host.present:
    - _connection_user: XXXXXXXXXX
    - _connection_password: XXXXXXXXXX
    - _connection_url: http://XXXXXXXXX/zabbix/api_jsonrpc.php
    - host: bat.mylab
    - visible_name: Cobra BAT
    - require_in:
      - assign-template-host-bat.mylab
    - groups:
      - Linux Desktops
    - interfaces:
      - 127.0.0.1_agent:
        - interface1: None
        - ip: 127.0.0.1
        - dns: bat.mylab
        - useip: False
        - type: agent
        - port: 10050
```

**Steps to Reproduce the behavior**
Write a state that creates a new host on Zabbix with a visible_name. The new host will be created, but the visible_name will be empty.

**Expected behavior**
It's expected to create a host in Zabbix with the "Visible Name" set as specified in state.

**Versions Report**
<details><summary>salt --versions-report</summary>
```
Salt Version:
           Salt: 3001
 
Dependency Versions:
           cffi: 1.14.0
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
         Jinja2: 2.11.2
        libgit2: Not Installed
       M2Crypto: 0.36.0
           Mako: 1.1.3
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: 2.20
       pycrypto: 2.6.1
   pycryptodome: 3.9.7
         pygit2: Not Installed
         Python: 3.8.6 (default, Sep 24 2020, 18:13:36)
   python-gnupg: Not Installed
         PyYAML: 5.3.1
          PyZMQ: 19.0.1
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.2
 
System Versions:
           dist: slackware 14.2 current
         locale: utf-8
        machine: x86_64
        release: 5.4.66
         system: Linux
        version: Slackware 14.2 current
```
</details>

### LINKED ISSUE #58603: [BUG] zabbix_host.present can't update an existing host
**Description**
When applying zabbix_host.present state to an existent host on Zabbix it fails by numerous errors:

- With Zabbix >= 5.0, the element "bulk" is not present in the hostinterfaces_get output. So it fails when "popped"
- When current and new inventory were missing, the state fails trying to iterate a boolean
- The update_interfaces try to delete all interfaces and then recreate them. That doesn't works as any interface with assigned items can't be deleted. Also, every host needs at least one default interface.

**Setup**
Slackware machines running Salt 3001 managing hosts in a Zabbix 5.0 server, also running on Slackware. The rendered state being applied is:
```
create-zabbix-host-bat.mylab:
  zabbix_host.present:
    - _connection_user: XXXXXXXXXX
    - _connection_password: XXXXXXXXXX
    - _connection_url: http://XXXXXXXXX/zabbix/api_jsonrpc.php
    - host: bat.mylab
    - visible_name: Cobra BAT
    - require_in:
      - assign-template-host-bat.mylab
    - groups:
      - Linux Desktops
    - interfaces:
      - 127.0.0.1_agent:
        - interface1: None
        - ip: 127.0.0.1
        - dns: bat.mylab
        - useip: False
        - type: agent
        - port: 10050
```

**Steps to Reproduce the behavior**
Write a state using zabbix_host.present that updates an existing host on Zabbix.

- inventory error
```
2020-09-16 17:37:00,981 [salt.state                                             :328 ][ERROR   ][32427] An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib64/python3.8/site-packages/salt/state.py", line 2153, in call
    ret = self.states[cdata["full"]](
  File "/usr/lib64/python3.8/site-packages/salt/loader.py", line 2087, in wrapper
    return f(*args, **kwargs)
  File "/usr/lib64/python3.8/site-packages/salt/states/zabbix_host.py", line 290, in present
    elif set(cur_inventory) != set(new_inventory):
TypeError: 'bool' object is not iterable
```
- "bulk" missing element
```
2020-09-15 20:21:53,504 [salt.state       :328 ][ERROR   ][13162] An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib64/python3.8/site-packages/salt/state.py", line 2153, in call
    ret = self.states[cdata["full"]](
  File "/usr/lib64/python3.8/site-packages/salt/loader.py", line 2087, in wrapper
    return f(*args, **kwargs)
  File "/usr/lib64/python3.8/site-packages/salt/states/zabbix_host.py", line 260, in present
    hostintf.pop("bulk")
KeyError: 'bulk'
```
- delete an interface with items linked
```
2020-09-28 10:56:36,872 [salt.state                                             :328 ][ERROR   ][30648] An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib64/python3.8/site-packages/salt/state.py", line 2153, in call
    ret = self.states[cdata["full"]](
  File "/usr/lib64/python3.8/site-packages/salt/loader.py", line 2087, in wrapper
    return f(*args, **kwargs)
  File "/usr/lib64/python3.8/site-packages/salt/states/zabbix_host.py", line 350, in present
    __salt__["zabbix.hostinterface_delete"](
  File "/usr/lib64/python3.8/site-packages/salt/modules/zabbix.py", line 1933, in hostinterface_delete
    ret = _query(method, params, conn_args["url"], conn_args["auth"])
  File "/usr/lib64/python3.8/site-packages/salt/modules/zabbix.py", line 199, in _query
    raise SaltException(
salt.exceptions.SaltException: Zabbix API: Invalid params. (Interface is linked to item "Host name of Zabbix agent running" on "Cobra BAT".)
```

**Expected behavior**
To apply the state in an already existing host in Zabbix and having the entry updated to the state specified.

**Versions Report**
<details><summary>salt --versions-report</summary>
```
Salt Version:
           Salt: 3001
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
         Jinja2: 2.11.2
        libgit2: Not Installed
       M2Crypto: 0.36.0
           Mako: 1.1.3
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: 3.9.7
         pygit2: Not Installed
         Python: 3.8.6 (default, Sep 24 2020, 18:13:36)
   python-gnupg: Not Installed
         PyYAML: 3.13
          PyZMQ: 19.0.1
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.2
 
System Versions:
           dist: slackware 14.2 current
         locale: utf-8
        machine: x86_64
        release: 5.4.66
         system: Linux
        version: Slackware 14.2 current
```
</details>

### LINKED ISSUE #58620: [BUG]: zabbix_host.present fails to add a SNMP interface in Zabbix 5.0
**Description**
zabbix_host.present fails to create a SNMP interface in Zabbix 5.0 because it requires new mandatory parameters. From [Zabbix documentation](https://www.zabbix.com/documentation/current/manual/api/reference/hostinterface/object):

| Property | Type  | Description                                                             |
|----------|-------|-------------------------------------------------------------------------|
| details  | array | Additional object for interface. *Required* if interface 'type' is SNMP.|

Inside "details"  are other mandatory properties, depending on SNMP version and security level.

**Setup**
Slackware machines running Salt 3001 managing hosts in a Zabbix 5.0 server, also running on Slackware. The rendered state being applied is:
```
create-zabbix-host-longshot:
  zabbix_host.present:
    - _connection_user: XXXXXXXXXX
    - _connection_password: XXXXXXXXXXX
    - _connection_url: http://XXXXXXXXXX/zabbix/api_jsonrpc.php
    - host: longshot.mylab
    - visible_name: "Longshot"
    - description: "Imaginary Friend"
    - groups:
        - "Network Devices"
    - interfaces:
        - interface1:
          - ip: 127.0.0.1
          - dns: longshot.mylab
          - useip: false
          - type: snmp
```
The error is:
```
[ERROR   ] An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib64/python3.8/site-packages/salt/state.py", line 2153, in call
    ret = self.states[cdata["full"]](
  File "/usr/lib64/python3.8/site-packages/salt/loader.py", line 2087, in wrapper
    return f(*args, **kwargs)
  File "/usr/lib64/python3.8/site-packages/salt/states/zabbix_host.py", line 369, in present
    host_create = __salt__["zabbix.host_create"](
  File "/usr/lib64/python3.8/site-packages/salt/modules/zabbix.py", line 1176, in host_create
    ret = _query(method, params, conn_args["url"], conn_args["auth"])
  File "/usr/lib64/python3.8/site-packages/salt/modules/zabbix.py", line 190, in _query
    raise SaltException(
salt.exceptions.SaltException: Zabbix API: Invalid params. (Incorrect arguments passed to function.)
```

**Steps to Reproduce the behavior**
Try to create or update a host with a SNMP interface in Zabbix 5.0 with zabbix_host.present state.

**Expected behavior**
It's expected to create the SNMP interface without major issues.

**Versions Report**
<details><summary>salt --versions-report</summary>
```
Salt Version:
           Salt: 3001
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
         Jinja2: 2.11.2
        libgit2: Not Installed
       M2Crypto: 0.36.0
           Mako: 1.1.3
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: 3.9.7
         pygit2: Not Installed
         Python: 3.8.6 (default, Sep 24 2020, 18:13:36)
   python-gnupg: Not Installed
         PyYAML: 3.13
          PyZMQ: 19.0.1
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.2
 
System Versions:
           dist: slackware 14.2 current
         locale: utf-8
        machine: x86_64
        release: 5.4.66
         system: Linux
        version: Slackware 14.2 current
```
</details>

### PR DESCRIPTION
**Pull request by piterpunk**
### What does this PR do?
This PR fixes allows zabbix_host.present to:

- Create a new host in Zabbix with a "Visible Name" configured, as was stated in state documentation.
- Compare and update an existing host
- Really update the existing interfaces instead trying removing and recreating them all.
- Added the mandatory parameters to create a SNMP hostinterface in Zabbix 5.0
- Now processes additional host properties like "description" or "inventory_mode".

### What issues does this PR fix or reference?
Fixes #58602 
Fixes #58603
Fixes #53838
Fixes #58620

### Previous Behavior
Host's "Visible Name" never was configured because zabbix_host.present didn't pass the parameter visible_name to zabbix.create_host. Other additional host properties were also ignored.

Host comparison and update fails because non existent key "bulk" and a failure in inventory check.

The update_interfaces didn't really update them, it try to removes and recreates them. Usually this strategy fails, as interfaces with linked items can't be removed. Also, no SNMP interface can be created in Zabbix 5.0

### New Behavior
Host's "Visible Name" is configured and additional host properties are processed. Host comparison and update runs Ok.

The update_interfaces now try to match the interface types and update them in place, then create the new ones and remove the rest.

SNMP interfaces can be created and updated seamlessly.

### Merge requirements satisfied?
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by piterpunk** [comment]
Converted to Work-In-Progress, I will include the fix for #53838 to keep the state 100% Ok with the documentation.

File: salt/states/zabbix_host.py [hunk 1]
```diff
@@ -224,17 +250,53 @@ def _interface_format(interfaces_data):
...
+            "ipmi_privilege",
+            "ipmi_username",
+            "status",
+            "tls_connect",
+            "tls_accept",
+            "tls_issuer",
+            "tls_subject",
+            "tls_psk_identity",
+            "tls_psk",
+        ]
```
Comment (OrangeDog) [inline salt/states/zabbix_host.py:273]: IMO this list should be unrestricted. Everything remaining in kwargs should be passed in the host definition. This allows using new Zabbix properties without having to wait for an update to Salt.

**Review by OrangeDog — CHANGES_REQUESTED** [review]

File: salt/states/zabbix_host.py (hunk 1 above)
Comment (piterpunk) [reply salt/states/zabbix_host.py:273]: At first I thought the same and just forwarded the remaining kwargs. But while coding I changed my mind for two reasons:

1. Some host properties needs to be adjusted before forwarding, like the existent "proxy_host" and "inventory". I guess it's better to have a new non-mandatory option ignored than sent wrongly.
2. The host properties set is very stable. From Zabbix 3.0 (2016) to 5.0 (2020) only the "tags" property was added. And this one is just the kind of property that I described before.

I still don't know the best way to handle the "tags" object, so I leave it out of this allowed list, until I figure out a good way. But all the other writable properties are supported and I can create and update a host on Zabbix 5.0.

Now I am writing the tests and fixing some bugs that the tests found, like failing when a host didn't have any additional property.

**Comment by piterpunk** [comment]
I fixed the last failing tests today and will be great to have this approved before began to test and adjust the other Zabbix related states for Zabbix 5+

**Review by garethgreenaway — DISMISSED** [review]

File: salt/states/zabbix_host.py [hunk 2]
```diff
@@ -224,17 +250,53 @@ def _interface_format(interfaces_data):
         for k, v in inv_item.items():
             new_inventory[k] = str(v)
 
+    if "visible_name" not in kwargs:
+        visible_name = None
+    else:
+        visible_name = kwargs.pop("visible_name")
```
Comment (twangboy) [inline salt/states/zabbix_host.py:256]: `visible_name = kwargs.pop("visible_name", None)`

**Review by twangboy — CHANGES_REQUESTED** [review]

File: salt/states/zabbix_host.py (hunk 2 above)
Comment (piterpunk) [reply salt/states/zabbix_host.py]: Thank you! This change makes the code easier to follow.

File: salt/states/zabbix_host.py [hunk 3]
```diff
@@ -182,40 +210,35 @@ def _interface_format(interfaces_data):
...
             except TypeError:
-                ret["comment"] = "Invalid proxy_host {0}".format(kwargs["proxy_host"])
+                ret["comment"] = "Invalid proxy_host {}".format(proxy_host)
                 return ret
 
-    if "inventory" not in kwargs:
-        inventory = {}
-    else:
-        inventory = kwargs["inventory"]
+    inventory = kwargs.pop("inventory", None)
```
Comment (twangboy) [inline salt/states/zabbix_host.py:241]: I think in this case it would need to be `kwargs.pop("inventory", {})`

**Review by twangboy — CHANGES_REQUESTED** [review]

File: salt/states/zabbix_host.py (hunk 3 above)
Comment (piterpunk) [reply salt/states/zabbix_host.py:241]: In the original code there is a safeguard if `inventory` is passed as `None`. Don't know the use case for this, probably formulas that always pass the `inventory` argument but with `None` when it is not needed. I try to not change the existent parsing.

If the user pass the `inventory = None`, the `kwargs.pop("inventory", {})` will pass the same `None` to the next steps and the `new_inventory` population will fail. That´s why I use `kwargs.pop("inventory", None)` and kept the conversion from `None` to `{}` in next line.

Comment (twangboy) [reply salt/states/zabbix_host.py:241]: Sounds good.

**Review by twangboy — APPROVED** [review]

**Comment by piterpunk** [comment]
@sagetherage to me it's important to have this PR merged before the code-freeze, I am using it at other projects and give maintenance to out-of-tree code isn't good.

**Comment by sagetherage** [comment]
@piterpunk we have it marked for `Aluminium` and code freeze is moved by a week and this is looking good! it needs to be rebased with master, but currently passing all tests - hurray!

**Review by garethgreenaway — APPROVED** [review]

**Review by Ch3LL — APPROVED** [review]

**Pull request merged**